ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/util/DataColumnSidecarELRecoveryManagerImpl.java

```diff
@@ -0,0 +1,371 @@
+  private final Spec spec;
+  private final AsyncRunner asyncRunner;
+  private final RecentChainData recentChainData;
+  private final ExecutionLayerChannel executionLayer;
+  private final int maxTrackers;
+  private final Consumer<List<DataColumnSidecar>> dataColumnSidecarPublisher;
+  private final CustodyGroupCountManager custodyGroupCountManager;
+  private final KZG kzg;
+
+  private final Supplier<MiscHelpersFulu> miscHelpersFuluSupplier;
+
+  @VisibleForTesting
+  public DataColumnSidecarELRecoveryManagerImpl(
+      final Spec spec,
+      final AsyncRunner asyncRunner,
+      final RecentChainData recentChainData,
+      final ExecutionLayerChannel executionLayer,
+      final UInt64 historicalSlotTolerance,
+      final UInt64 futureSlotTolerance,
+      final int maxTrackers,
+      final KZG kzg,
+      final Consumer<List<DataColumnSidecar>> dataColumnSidecarPublisher,
+      final CustodyGroupCountManager custodyGroupCountManager) {
+    super(spec, futureSlotTolerance, historicalSlotTolerance);
+    this.spec = spec;
+    this.asyncRunner = asyncRunner;
+    this.recentChainData = recentChainData;
+    this.executionLayer = executionLayer;
+    this.maxTrackers = maxTrackers;
+    this.kzg = kzg;
+    this.dataColumnSidecarPublisher = dataColumnSidecarPublisher;
+    this.custodyGroupCountManager = custodyGroupCountManager;
+    this.miscHelpersFuluSupplier =
+        () -> MiscHelpersFulu.required(spec.forMilestone(SpecMilestone.FULU).miscHelpers());
+  }
+
+  @Override
+  public void onNewDataColumnSidecar(
+      final DataColumnSidecar dataColumnSidecar, final RemoteOrigin remoteOrigin) {
+    if (!spec.atSlot(dataColumnSidecar.getSlot())
+        .getMilestone()
+        .isGreaterThanOrEqualTo(SpecMilestone.FULU)) {
+      return;
+    }
+    if (recentChainData.containsBlock(dataColumnSidecar.getBlockRoot())) {
+      return;
+    }
+    if (shouldIgnoreItemAtSlot(dataColumnSidecar.getSlot())) {
+      return;
+    }
+
+    final SlotAndBlockRoot slotAndBlockRoot = dataColumnSidecar.getSlotAndBlockRoot();
+
+    if (createRecoveryTaskFromDataColumnSidecar(dataColumnSidecar)) {
+      onFirstSeen(slotAndBlockRoot, Optional.of(remoteOrigin));
+      orderedRecoveryTasks.add(slotAndBlockRoot);
+    }
+  }
+
+  private synchronized boolean createRecoveryTaskFromDataColumnSidecar(
+      final DataColumnSidecar dataColumnSidecar) {
+    if (recoveryTasks.containsKey(dataColumnSidecar.getBlockRoot())) {
+      return false;
+    }
+    if (!dataColumnSidecar.getSlot().equals(getCurrentSlot())) {
+      return false;
+    }
+    makeRoomForNewTracker();
+    recoveryTasks.put(
+        dataColumnSidecar.getBlockRoot(),
+        new RecoveryTask(
+            dataColumnSidecar.getSignedBeaconBlockHeader(),
+            dataColumnSidecar.getSszKZGCommitments(),
+            dataColumnSidecar.getKzgCommitmentsInclusionProof().asListUnboxed()));
+    return true;
+  }
```

Comment:
We might need to review some of this synchronous access here. I believe move of these checks could be done w/o locking (e.g. recoveryTasks.containsKey). Potentially `recoveryTasks` can be a concurrency-friendly structure and we can avoid most of the synchronized logic. What do you think?